### PR TITLE
fix(drizzle): warn on truncated identifiers, throw on collisions

### PR DESCRIPTION
**Pull request by PatrikKozak**
Surfaces Postgres identifier truncation instead of letting it silently desync the schema.

Postgres truncates any identifier over 63 characters. Payload builds column and inline index/foreign-key names by concatenating nested field paths, so deeply nested fields can exceed the limit. Postgres then stores a shorter name than Payload's schema and migration snapshot expect, and dev `push` hangs on a repeated `is <column> created or renamed?` prompt on every boot.

#17563 addressed this for table names by throwing, but that only reaches base and companion tables. Columns and inline index/foreign-key names were still truncated silently, and throwing outright blocks existing installs that already carry a truncated name from upgrading.

Adds `checkTruncatedIdentifiers`, run during postgres `init` after `beforeSchemaInit`:

- Warns (dev only) when a single identifier exceeds 63 characters, so existing installs keep booting.
- Throws when two identifiers truncate to the same name, since Postgres cannot create that schema anyway — and a colliding schema could never have booted, so this never breaks a running app.

Base table and enum names still throw during construction via `validateIdentifierLength`, unchanged.

Related: softens the companion-table checks from #17563 (`_locales`, `_rels`, `_texts`, `_numbers`) from a hard throw to the same warn-or-collision handling.

### CODE REVIEW

**Comment by DanRibbens** [comment]
Overrides https://github.com/payloadcms/payload/pull/14637

File: packages/drizzle/src/utilities/checkTruncatedIdentifiers.ts [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+      for (const indexKey in table.indexes) {
+        check(table.indexes[indexKey].name, indexNames)
+      }
+    }
+
+    if (table.foreignKeys) {
+      for (const foreignKeyKey in table.foreignKeys) {
+        check(table.foreignKeys[foreignKeyKey].name, foreignKeyNames)
+      }
+    }
```
Comment (DanRibbens) [inline packages/drizzle/src/utilities/checkTruncatedIdentifiers.ts:69]: These two branches (index names and FK names) are effectively dead code and duplicate logic that already lives in `buildIndexName`/`buildForeignKeyName`.

Every `indexes[...].name` and `foreignKeys[...].name` in the raw schema is produced by those builders — nothing assigns an index/FK name raw. And both builders already:

1. **Truncate** to ≤ 60 chars (`if (indexName.length > 60) ...slice(0, 60 - suffix.length)`), and
2. **Dedupe** by recursing with an incrementing `_N` suffix until the name is unique in `adapter.indexes` / `adapter.foreignKeys`.

So by the time we read these names here, they're always ≤ 60 and already collision-free. `truncate()` only slices when `length > 63`, so `final === original` every time — nothing is ever added to `truncatedNames`, and no collision can register. The `indexNames`/`foreignKeyNames` maps and both loops can never fire (the spec only exercises column and table collisions, never index/FK).

The reason is architectural, and it's the same reason the check *is* needed for columns: `buildIndexName` makes Payload use the *same* truncated name it hands Postgres, so there's no desync to detect. The desync bug only exists for identifiers that aren't run through a truncating builder — i.e. table names and column names, which are the two namespaces worth keeping here.

Suggest dropping the `indexNames`/`foreignKeyNames` maps and these two loops and keeping only the table + column checks. (If you'd rather this function be the single source of truth, the alternative is to stop truncating in the builders and let this own all four namespaces — but that forfeits their auto-dedup, so I'd keep the builders and trim here.)

File: packages/drizzle/src/utilities/checkTruncatedIdentifiers.ts [hunk 2]
```diff
@@ -0,0 +1,87 @@
+import { APIError } from 'payload'
+
+import type { DrizzleAdapter } from '../types.js'
+
+import { maxIdentifierLength } from './validateIdentifierLength.js'
+
+const truncate = (name: string): string =>
+  name.length > maxIdentifierLength ? name.slice(0, maxIdentifierLength) : name
```
Comment (DanRibbens) [inline packages/drizzle/src/utilities/checkTruncatedIdentifiers.ts:8]: Minor consistency nit: this uses `maxIdentifierLength` (63) while `buildIndexName`/`buildForeignKeyName` cap at a hardcoded **60**. They're intentionally different (60 leaves headroom for the `_idx`/`_fk`/`_N` suffixes), but nothing documents that the two limits are related, so a future edit could drift them apart. Consider a shared named constant — e.g. deriving the 60 as `maxIdentifierLength - maxSuffixLength` — to make the relationship explicit.

Comment (PatrikKozak) [reply packages/drizzle/src/utilities/checkTruncatedIdentifiers.ts:8]: Good call - done. 

Added `maxGeneratedIdentifierLength = maxIdentifierLength - 3` next to `maxIdentifierLength` and swapped both builders to use it, so the relationship is explicit and the two limits can't drift apart. `truncate()` here stays at `maxIdentifierLength` (63) since it's detecting Postgres's actual limit.

File: packages/drizzle/src/utilities/checkTruncatedIdentifiers.ts (hunk 1 above)
Comment (PatrikKozak) [reply packages/drizzle/src/utilities/checkTruncatedIdentifiers.ts:69]: Yeah, agreed on the builder names - anything going through `buildIndexName` / `buildForeignKeyName` is already truncated and deduped, so the check does nothing for those. No argument there.

The catch is that not everything goes through the builders. The structural index / FK names for `array` / `blocks` / `select` sub-tables get built inline in `schema/traverseFields.ts` - stuff like `${arrayTableName}_order_idx` and `${arrayTableName}_parent_id_fk` (around lines 205/216, 444–480, 826–870) - and buildTable drops those straight into `baseIndexes` / `baseForeignKeys` as-is. 

They never hit the builders, so nothing pre-truncates them.

The sub-table name itself is validated ≤ 63, but tacking on `_parent_id_fk` (+14) can blow past it. I checked with a quick test - collection `p` + a 58-char array field gives array table `p_<58>` (60 chars, fine), but its inline `_order_idx` (70) and `_parent_id_fk` (73) both overflow and get flagged. If they all went through the builders, nothing would've flagged here.

So I've kept the index / FK namespaces and added the missing coverage (an end-to-end test for the inline overflow, plus unit tests for inline index/FK collisions) - the spec gap was the real issue, not the branches.

Happy to instead route those inline sites through the builders and then drop these branches, which matches the "single truncating source" framing - but that changes generated index / FK names for existing installs (migration churn), so I'd do it as a separate follow-up rather than fold it in here. Let me know which you'd prefer.

Comment (DanRibbens) [reply packages/drizzle/src/utilities/checkTruncatedIdentifiers.ts:69]: Let's avoid dropping these branches considering we want to avoid generating migrations for the indexes + FKs. I agree that is not something we want to take on this PR.

**Review by DanRibbens — APPROVED** [review]
